src/memray/_memray.pyx

```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
+
+
+def compute_statistics(
+    file_name,
+    *,
+    report_progress=False,
+    num_largest=5,
+):
+    cdef shared_ptr[RecordReader] reader_sp = make_shared[RecordReader](
+        unique_ptr[FileSource](new FileSource(file_name))
+    )
+    cdef RecordReader* reader = reader_sp.get()
+
+    cdef header = reader.getHeader()
+    total = header["stats"]["n_allocations"] or None
+
+    cdef AllocationStatsAggregator aggregator
+    cdef ProgressIndicator progress_indicator = ProgressIndicator(
+        "Computing statistics",
+        total=total,
+        report_progress=report_progress,
+    )
+    with progress_indicator:
+        while True:
+            PyErr_CheckSignals()
+            ret = reader.nextRecord()
+            if ret == RecordResult.RecordResultAllocationRecord:
+                aggregator.addAllocation(reader.getLatestAllocation())
+                progress_indicator.update(1)
+            elif ret == RecordResult.RecordResultMemoryRecord:
+                pass
+            else:
+                break
 
-        stats = self._header["stats"]
-        allocator_id_to_name = {
-            PythonAllocatorType.PYTHON_ALLOCATOR_PYMALLOC: "pymalloc",
-            PythonAllocatorType.PYTHON_ALLOCATOR_PYMALLOC_DEBUG: "pymalloc debug",
-            PythonAllocatorType.PYTHON_ALLOCATOR_MALLOC: "malloc",
-            PythonAllocatorType.PYTHON_ALLOCATOR_OTHER: "unknown",
-        }
-        python_allocator = allocator_id_to_name[self._header["python_allocator"]]
-        return Metadata(start_time=millis_to_dt(stats["start_time"]),
-                        end_time=millis_to_dt(stats["end_time"]),
-                        total_allocations=stats["n_allocations"],
-                        total_frames=stats["n_frames"],
-                        peak_memory=self._high_watermark.peak_memory,
-                        command_line=self._header["command_line"],
-                        pid=self._header["pid"],
-                        python_allocator=python_allocator,
-                        has_native_traces=self._header["native_traces"])
+    # Ignore the n_allocations in the header, use our observed value.
+    header["stats"]["n_allocations"] = progress_indicator.num_processed
+
+    # NOTE: We need to cast away the constness of some references below,
+    #       because Cython's iterator handling is not const-correct.
+
+    # Convert allocation counts by allocator/by size to Python dicts.
+    allocation_count_by_allocator = {
+        AllocatorType(<int>id_and_count.first).name: id_and_count.second
+        for id_and_count in <unordered_map[Allocator, uint64_t]&>(
+            aggregator.allocationCountByAllocator()
+        )
+    }
+    allocation_count_by_size = {
+        size_and_count.first: size_and_count.second
+        for size_and_count in <unordered_map[size_t, uint64_t]&>(
+            aggregator.allocationCountBySize()
+        )
+    }
```

Comment:
```suggestion
    allocation_count_by_size = aggregator.allocationCountBySize()
```

Reply:
Same as before, Cython should be able to return this automatically, no?

Reply:
I couldn't find the right magical incantation to make that happen, but Pablo and I played with it offline and got this working.